lib/trie/trie_test.go

```diff
@@ -62,46 +79,734 @@ const (
+			string([]byte{0xa0, 0x2b}): []byte("leaf"),
+			string([]byte{0x0a, 0x2b}): []byte("leaf"),
+		}
+
+		entriesMatch(t, expectedEntries, entries)
+	})
+
+	t.Run("custom root", func(t *testing.T) {
+		t.Parallel()
+
+		root := &node.Branch{
+			Key:   []byte{0xa, 0xb},
+			Value: []byte("root"),
+			Children: [16]node.Node{
+				nil, nil, nil,
+				&node.Branch{ // branch with value at child index 3
+					Key:   []byte{0xb},
+					Value: []byte("branch 1"),
+					Children: [16]node.Node{
+						nil, nil, nil,
+						&node.Leaf{ // leaf at child index 3
+							Key:   []byte{0xc},
+							Value: []byte("bottom leaf"),
+						},
+					},
+				},
+				nil, nil, nil,
+				&node.Leaf{ // leaf at child index 7
+					Key:   []byte{0xd},
+					Value: []byte("top leaf"),
+				},
+				nil,
+				&node.Branch{ // branch without value at child index 9
+					Key: []byte{0xe},
+					Children: [16]node.Node{
+						nil, nil, nil,
+					},
+				},
+			},
+		}
+
+		trie := NewTrie(root)
+
+		entries := trie.Entries()
+
+		expectedEntries := map[string][]byte{
+			string([]byte{0xab}):             []byte("root"),
+			string([]byte{0xab, 0x3b}):       []byte("branch 1"),
+			string([]byte{0xab, 0x7d}):       []byte("top leaf"),
+			string([]byte{0xab, 0x3b, 0x3c}): []byte("bottom leaf"),
+		}
+
+		entriesMatch(t, expectedEntries, entries)
+	})
+
+	t.Run("end to end", func(t *testing.T) {
+		t.Parallel()
+
+		trie := &Trie{
+			root:        nil,
+			childTries:  make(map[common.Hash]*Trie),
+			deletedKeys: make([]common.Hash, 0),
+		}
+
+		kv := map[string][]byte{
+			"ab":  []byte("pen"),
+			"abc": []byte("penguin"),
+			"hy":  []byte("feather"),
+			"cb":  []byte("noot"),
+		}
+
+		for k, v := range kv {
+			trie.Put([]byte(k), v)
+		}
+
+		entries := trie.Entries()
+
+		assert.Equal(t, kv, entries)
+	})
+}
+
+func Test_NextKey(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		trie    *Trie
+		key     []byte
+		nextKey []byte
+	}{
+		"nil root and nil key returns nil": {
+			trie: &Trie{},
+		},
+		"nil root returns nil": {
+			key:  []byte{1},
+			trie: &Trie{},
+		},
+		"nil key returns root leaf": {
+			trie: &Trie{
+				root: &node.Leaf{
+					Key: []byte{1},
+				},
+			},
+			nextKey: []byte{1}, // TODO should this be nil?
+		},
+		"nil key returns root branch": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{1},
+				},
+			},
+			// nextKey: []byte{1}, TODO it returns nil for now
+		},
+		"key greater than root branch full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x01},
+				},
+			},
+			key: []byte{0x02},
+		},
+		"key equal to root branch full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x01},
+				},
+			},
+			key: []byte{0x01},
+		},
+		"key smaller than root branch full key": {
+			trie: &Trie{
+				root: &node.Branch{
+					Key: []byte{0x02},
+				},
+			},
+			key: []byte{0x01},
+			// nextKey: []byte{0x02}, TODO it returns nil for now
```

Comment:
I suspect this is because the node key is in nibbles form, while  the input to NextKey is ib bytes form, so you're comparing `nibble(2)` with `[nibble(0), nibble(1)]`, so the next key is nil

Reply:
I think it's just because the root is a branch without a value so that's an invalid trie as well. Changing the branch to a leaf fixed it. Thanks!!